src/coupling.jl

```diff
@@ -239,3 +240,157 @@ function apply_coupling!(sys::System{D, G, T}, barostat::MonteCarloBarostat, sim
+If `ΔW ≤ 0` the step is always accepted, if `ΔW > 0` the step is accepted with probability
+`exp(-ΔW/kT)`.
+
+The scale factor is modified over time to maintain an acceptance rate of around half.
+If the topology of the system is set then molecules are moved as a unit so properties
+such as bond lengths do not change.
+
+The barostat assumes that the simulation is being run at a constant temperature but
+does not actively control the temperature.
+It should be used alongside a temperature coupling method such as the [`Langevin`](@ref)
+simulator or [`AndersenThermostat`](@ref) coupling.
+The neighbor list is not updated when making trial moves or after accepted moves.
+Note that the barostat can change the bounding box of the system.
+For 3D systems, `pressure` is a SVector of lenght 3 with components pressX, pressY,
+and pressZ representing the target pressure in each axis.
+For 2D systems, `pressure` is a SVector of lenght 2 with components pressX and pressY.
+To keep an axis fixed, set the corresponding pressure to `nothing`.
+
+Not currently compatible with automatic differentiation using Zygote.
+"""
+mutable struct MonteCarloAnisotropicBarostat{D, T, P, K, V}
+    pressure::SVector{D,P}
+    temperature::K
+    n_steps::Int
+    n_iterations::Int
+    volume_scale::V
+    scale_increment::T
+    max_volume_frac::T
+    trial_find_neighbors::Bool
+    n_attempted::Vector{Int}
+    n_accepted::Vector{Int}
+end
+
+function MonteCarloAnisotropicBarostat(
+    pressure::SVector{D},
+    temperature,
+    boundary;
+    n_steps=30,
+    n_iterations=1,
+    scale_factor=0.01,
+    scale_increment=1.1,
+    max_volume_frac=0.3,
+    trial_find_neighbors=false,
+) where {D}
+    volume_scale = box_volume(boundary) * float_type(boundary)(scale_factor)
+    volume_scale = fill(volume_scale, D)
+    length(boundary.side_lengths) != D && error("pressure is not compatible with boundary")
+
+    return MonteCarloAnisotropicBarostat(
+        pressure,
+        temperature,
+        n_steps,
+        n_iterations,
+        volume_scale,
+        scale_increment,
+        max_volume_frac,
+        trial_find_neighbors,
+        fill(0, D),
+        fill(0, D),
+    )
+end
+
+function apply_coupling!(
+    sys::System{D, G, T},
+    barostat::MonteCarloAnisotropicBarostat{D},
+    sim,
+    neighbors=nothing,
+    step_n::Integer=0;
+    n_threads::Integer=Threads.nthreads()
+) where {D, G, T}
+
+    !iszero(step_n % barostat.n_steps) && return false
+    all(isnothing, barostat.pressure) && return false
+
+    kT = sys.k * barostat.temperature
+    n_molecules = isnothing(sys.topology) ? length(sys) : length(sys.topology.molecule_atom_counts)
+    recompute_forces = false
+
+    axis = undef
+    while true
+        axis = rand(1:D)
+        !isnothing(barostat.pressure[axis]) && break
+    end
+    mask1 = falses(D)
+    mask2 = trues(D)
+    mask1[axis] = true
+    mask2[axis] = false
+
+    for attempt_n in 1:barostat.n_iterations
```

Comment:
The `n_iterations` parameter is not in the OpenMM implementation but I added it in case people want to run multiple trials per step. Given that these trials should be independent, I think the above lines picking the dimension should go inside the loop.

Reply:
I moved the axis selection part inside the trial loop:

```julia
for attempt_n in 1:barostat.n_iterations
        axis = 0
        while true
            axis = rand(1:D)
            !isnothing(barostat.pressure[axis]) && break
        end
        mask1 = falses(D)
        mask2 = trues(D)
        mask1[axis] = true
        mask2[axis] = false
```